src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -252,52 +257,114 @@ public static CachingContext GetOrCreate(JsonSerializerOptions options)
         /// If two instances are equivalent, they should generate identical metadata caches;
         /// the converse however does not necessarily hold.
         /// </summary>
-        private static bool AreEquivalentOptions(JsonSerializerOptions left, JsonSerializerOptions right)
+        private sealed class EqualityComparer : IEqualityComparer<JsonSerializerOptions>
         {
-            Debug.Assert(left != null && right != null);
-
-            return
-                left._dictionaryKeyPolicy == right._dictionaryKeyPolicy &&
-                left._jsonPropertyNamingPolicy == right._jsonPropertyNamingPolicy &&
-                left._readCommentHandling == right._readCommentHandling &&
-                left._referenceHandler == right._referenceHandler &&
-                left._encoder == right._encoder &&
-                left._defaultIgnoreCondition == right._defaultIgnoreCondition &&
-                left._numberHandling == right._numberHandling &&
-                left._unknownTypeHandling == right._unknownTypeHandling &&
-                left._defaultBufferSize == right._defaultBufferSize &&
-                left._maxDepth == right._maxDepth &&
-                left._allowTrailingCommas == right._allowTrailingCommas &&
-                left._ignoreNullValues == right._ignoreNullValues &&
-                left._ignoreReadOnlyProperties == right._ignoreReadOnlyProperties &&
-                left._ignoreReadonlyFields == right._ignoreReadonlyFields &&
-                left._includeFields == right._includeFields &&
-                left._propertyNameCaseInsensitive == right._propertyNameCaseInsensitive &&
-                left._writeIndented == right._writeIndented &&
-                left._typeInfoResolver == right._typeInfoResolver &&
-                CompareLists(left._converters, right._converters);
-
-            static bool CompareLists<TValue>(ConfigurationList<TValue> left, ConfigurationList<TValue> right)
+            public bool Equals(JsonSerializerOptions? left, JsonSerializerOptions? right)
+            {
+                Debug.Assert(left != null && right != null);
+
+                return
+                    left._dictionaryKeyPolicy == right._dictionaryKeyPolicy &&
```

Comment:
should you start with reference equality check here?

Reply:
Since the class isn't defining its own equality operator, this is using reference equality.

Reply:
It's not needed -- each options instance will insert itself at most once inside the shared cache. The same instance could only appear if multiple threads are racing when looking up the cache outside of the lock, but generally speaking that scenario shouldn't apply in cases where people don't reuse options instances.